### ISSUE TITLE
no-use-before-define - Add option for specific ignores (like useStyles)

### ISSUE THREAD

**Issue by jeffvandyke** [eslint/eslint#13422]
**What rule do you want to change?**
no-use-before-define - new option

**Does this change cause the rule to produce more or fewer warnings?**
Fewer, and the option would report fewer errors if specified.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option for ignoring user-specified variables like `useStyles`, to allow less-important styles to be relocated to the bottom, or potentially other cases where you might want to save uninteresting variables for last, as a pattern throughout an entire codebase.

**Please provide some example code that this change will affect:**

```js
import { makeStyles } from '@material-ui/core';

function MyComponent() {
  const classes = useStyles();
  return <div className={classes.myDiv}>Stuff</div>;
}

const useStyles = makeStyles({
  myDiv: {
    backgroundColor: '#ddd';
  }
});
```

**What does the rule currently do for this code?**
An error is reported because useStyles is defined after it's used, though in this context it's safe because React components are instantiated and run their code after the file has finished defining all variables.

**What will the rule do after it's changed?**
This code would not report an error - useStyles could be specified as an exception to the no-use-before-define

**Are you willing to submit a pull request to implement this change?**
...Quite possibly, though I only have minimal experience with eslint rule code.

**Comment by anikethsaha** [comment]
Thanks for the issue, 

IMO, this will be an error if es6 is concerned as this rule's scope is to protect the es6+ code. (irrespective of the env like react, node. )

maybe this is a good addition for [eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react) cause as per the description, this is particularly for the react environment.

**Comment by kaicataldo** [comment]
I believe the [`{ variables: false }`](https://eslint.org/docs/rules/no-use-before-define#options) configuration option is what you're looking for. If you have already tried it, can you explain why it didn't solve your problem? You can see it in action [here](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdXNlLWJlZm9yZS1kZWZpbmU6IFtcImVycm9yXCIsIHsgXCJ2YXJpYWJsZXNcIjogZmFsc2UgfV0qL1xuXG5pbXBvcnQgeyBtYWtlU3R5bGVzIH0gZnJvbSAnQG1hdGVyaWFsLXVpL2NvcmUnO1xuXG5mdW5jdGlvbiBNeUNvbXBvbmVudCgpIHtcbiAgY29uc3QgY2xhc3NlcyA9IHVzZVN0eWxlcygpO1xuICByZXR1cm4gPGRpdiBjbGFzc05hbWU9e2NsYXNzZXMubXlEaXZ9PlN0dWZmPC9kaXY+O1xufVxuXG5jb25zdCB1c2VTdHlsZXMgPSBtYWtlU3R5bGVzKHtcbiAgbXlEaXY6IHtcbiAgICBiYWNrZ3JvdW5kQ29sb3I6ICcjZGRkJ1xuICB9XG59KTsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

**Comment by jeffvandyke** [comment]
Ordinarily, I do want to make sure that variables are defined before they are used in this file, it's just the unique situation with `useStyles` that I want to ignore that rule for. My annoyance is having to scroll past `useStyles` in a react component file before I get to the code that I'm almost always interested in. I could put an `eslint-disable-next-line no-use-before-define` by every occurrence of `useStyles` in my project, but it'd be better to make it a project rule, and having an option for ignoring certain variables would allow people to do that. Does that make sense?

**Comment by kaicataldo** [comment]
Thanks for the explanation. This sounds to me like it falls under our new policy to freeze stylistic rules (please see https://eslint.org/blog/2020/05/changes-to-rules-policies for more details).

The good news is that this should be fairly easy to enforce by creating a custom rule using [eslint-rule-composer](https://www.npmjs.com/package/eslint-rule-composer)!

**Comment by jeffvandyke** [comment]
Thanks for informing me about that change, was not aware! I'll be sure to look into eslint-rule-composer, though that might be its own challenge :)
